src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -213,70 +201,18 @@ public void Remove(object key)
 
             CheckDisposed();
             CacheEntry entry;
-            _entryLock.EnterReadLock();
-            try
-            {
-                if (_entries.TryGetValue(key, out entry))
-                {
-                    entry.SetExpired(EvictionReason.Removed);
-                }
-            }
-            finally
-            {
-                _entryLock.ExitReadLock();
-            }
-
-            if (entry != null)
+            if (_entries.TryRemove(key, out entry))
             {
-                // TODO: For efficiency consider processing these removals in batches.
-                RemoveEntry(entry);
+                entry.SetExpired(EvictionReason.Removed);
+                entry.InvokeEvictionCallbacks();
             }
 
             StartScanForExpiredItems();
         }
 
         private void RemoveEntry(CacheEntry entry)
         {
-            _entryLock.EnterWriteLock();
-            try
-            {
-                // Only remove it if someone hasn't modified it since our lookup
-                CacheEntry currentEntry;
-                if (_entries.TryGetValue(entry.Key, out currentEntry)
-                    && object.ReferenceEquals(currentEntry, entry))
-                {
-                    _entries.Remove(entry.Key);
-                }
-            }
-            finally
-            {
-                _entryLock.ExitWriteLock();
-            }
-            entry.InvokeEvictionCallbacks();
-        }
-
-        private void RemoveEntries(List<CacheEntry> entries)
-        {
-            _entryLock.EnterWriteLock();
-            try
-            {
-                foreach (var entry in entries)
-                {
-                    // Only remove it if someone hasn't modified it since our lookup
-                    CacheEntry currentEntry;
-                    if (_entries.TryGetValue(entry.Key, out currentEntry)
-                        && object.ReferenceEquals(currentEntry, entry))
-                    {
-                        _entries.Remove(entry.Key);
-                    }
-                }
-            }
-            finally
-            {
-                _entryLock.ExitWriteLock();
-            }
-
-            foreach (var entry in entries)
+            if (_entriesCollection.Remove(new KeyValuePair<object, CacheEntry>(entry.Key, entry)))
```

Comment:
Because we want to make sure no updates to the entry has been made before we remove it.